MaibornWolff/codecharta
# Tech/rewrite file validation

# Tech/rewrite file validation

## Issue
As a developer I want to guarantee that the imported files contain the expected data right before pushing their file-content into the application state, in order to be able to use it appropriately afterwards.

## Current situation
CodeCharta uses a self written json Schema (find app/codeCharta/util/schema.json) to validate the incoming imported files. The issue with this is, the json schema is very difficult to read, hard to maintain, when the corresponding interface changes and therefore it's outdated and doesn't validate, what we actually want it to validate.

## Solution
Use [typescript-json-schema](https://github.com/YousefED/typescript-json-schema) to generate the json schema from the well defined interface `ExportCCFile` inside the `codeCharta.model.ts` file. This ensures that: 
- nobody needs to understand the internals of the json schema
- everybody is able to easily update the schema using the command `npm run schema:generate`
- The schema is up to date and correct IN DETAIL:
  - It reports errors with wrong enum parameters (like with `AttributeType`)
  - It knows which properties are required and which are optional from the interface `?` character.

Tech: The following command generates the file. It uses the following parameters:
- `--required` makes properties without the `?` required
- `--topRef` makes it easier to read the json schema with a top level reference, which references to the definition for the `ExportCCFile` schema, which then recursively calls other definitions.
```
typescript-json-schema ./app/codeCharta/codeCharta.model.ts ExportCCFile --required --topRef --out app/codeCharta/util/generatedSchema.json
```

Further necessary changes:
- update dataMocks and remove outdated properties like `fileName` and `id`
- update interface `ExportCCFile` to reflect the actual expected content with optional properties and the different BlacklistType `hide`

## How to test
You can use the online [JSON Schema validator](https://www.jsonschemavalidator.net), copy paste the generatedSchema and a cc.json of your choice. Then you can change the cc.json data and check if it reports the errors or not. Of course you can also import files into the running visualization of this branch, as it uses the new generatedSchema.

## TODO
- [x] Move HTML generation out of the dialogService (undo the changes to create a new dialog pop up)
- [x] Validate apiVersion and migrate afterwards (see [comment](https://github.com/MaibornWolff/codecharta/pull/860#issuecomment-586589507))
- [x] Fix resetting fileStates to early when importing invalid files, which results in a state without any files
- [x] Copy or generate the same validation file for analysis/tools/ValidationTool/src/main/resources/cc.json

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [ ] **All** requirements mentioned in the issue are implemented
- [ ] Does match the Code of Conduct and the Contribution file
- [ ] Task has its own **GitHub issue** (something it is solving)
  - [ ] Issue number is **included in the commit messages** for traceability
- [ ] **Update the README.md** with any changes/additions made
- [ ] **Update the CHANGELOG.md** with any changes/additions made
- [ ] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [ ] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [ ] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [ ] **Manual testing** did not fail
